<!doctype html>
<html lang="en">

<head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Promesas</title>
</head>

<body>
    <h1>Promesas</h1> 
    <script>    
        const irAlRestaurant = () => {
            return 'llegue al restaurant';
        }
    
        const pedirComida = () => {
            return new Promise((res, rej) => {
                setTimeout(() => {
                    if (false) {  
                        res('recibo comida');
                    } else {
                        rej('no hay comida');
                    }
                }, 2000);
            });
        }
    
        const pedirPostre = () => {
            return new Promise((res, rej) => {
                setTimeout(() => {
                    if (true) { 
                        res('recibo el postre');
                    } else {
                        rej('no hay postre');
                    }
                }, 2000);
            });
        }
    
        const voyAPagar = () => {
            return new Promise((res) => {
                setTimeout(() => {
                    res('pago');
                }, 2000);
            });
        }
    
        const casa = () => {
            return 'llegue a casa';
        }
    
        console.log(irAlRestaurant());
    
        pedirComida()
            .then(respuesta => {
                console.log(respuesta);
                return pedirPostre();
            })
            .then(respuesta => {
                console.log(respuesta);
                return voyAPagar();
            })
            .then(respuesta => {
                console.log(respuesta);
            })
            .catch(error => {
                console.error('Error:', error);
            })
            .finally(() => {
                console.log(casa());
            });
    </script> 

    <!-- <script>  este codigo funciona con un error en finally o .catch.
                    revisar. El codigo de arriba funciona.
        const irAlRestaurant = () => {
            return 'llegue al restaurant'
        }

        const pedirComida = () => {
            return new Promise((res, rej) => {
                setTimeout(() => {
                    if (false) {
                        res('recibo comida')
                    } else {
                        rej('no hay comida')
                    }
                }, 2000);
            })


        }

        const pedirPostre = () => {
            return new Promise((res, rej) => {
                setTimeout(() => {
                    if (true) {
                        res('recibo el postre')
                    } else {
                        rej('no hay postre')
                    }

                }, 2000);
            })


        }

        const voyAPagar = () => {
            return new Promise((res) => {
                setTimeout(() => {
                    res('pago')
                }, 2000);
            })


        }

        const casa = () => {
            return ("llegue a casa")
        }


        console.log(irAlRestaurant())

        pedirComida().then((respuesta) => {
            console.log(respuesta)
            pedirPostre().then((respuesta) => {
                console.log(respuesta)
                voyAPagar().then((respuesta) => {
                    console.log(respuesta)
                })
            }) .finally(() => {
                console.log(casa())
            })

        }).catch((error) => {
            console.log(error)
        }) 
    </script> -->
</body>

</html>